src/orange/detail/BIHTraversalHelper.hh

```diff
@@ -0,0 +1,154 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2022-2024 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file orange/detail/BIHTraversal.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include "../OrangeData.hh"
+
+namespace celeritas
+{
+namespace detail
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Traverse BIH tree using a depth-first search.
+ *
+ * \todo move to top-level orange directory out of detail namespace
+ */
+class BIHTraversalHelper
+{
+  public:
+    //!@{
+    //! \name Type aliases
+    using Storage = NativeCRef<BIHTreeData>;
+    //!@}
+
+    // Construct from vector of bounding boxes and storage for LocalVolumeIds
+    inline CELER_FUNCTION
+    BIHTraversalHelper(BIHTree const& tree, Storage const& storage);
+
+    // Determine if a node is inner, i.e., not a leaf
+    inline CELER_FUNCTION bool is_inner(BIHNodeId id) const;
+
+    // Get an inner node for a given BIHNodeId
+    inline CELER_FUNCTION BIHInnerNode const&
+    get_inner_node(BIHNodeId id) const;
+
+    // Get a leaf node for a given BIHNodeId
+    inline CELER_FUNCTION BIHLeafNode const& get_leaf_node(BIHNodeId id) const;
+
+    // Get a bbox for a given vol_id
+    inline CELER_FUNCTION FastBBox const& get_bbox(LocalVolumeId vol_id) const;
+
+    // Get the vol_id of the ith volume on a given leaf node
+    inline CELER_FUNCTION LocalVolumeId get_leaf_volid(BIHLeafNode leaf,
+                                                       size_type i) const;
+
+    // Get the number of inf_volids on the tree
+    inline CELER_FUNCTION size_type get_num_inf_volids() const;
+
+    // Get the ith vol_id in inf_volids
+    inline CELER_FUNCTION LocalVolumeId get_inf_volid(size_type i) const;
+
+  private:
+    //// DATA ////
+    BIHTree const& tree_;
+    Storage const& storage_;
+    size_type leaf_offset_;
```

Comment:
I don't think it's necessary to cache this (using another register to do so): just have a helper function `size_type leaf_offset() const`.